### PR TITLE
feat: async requests and load tests

### PR DESCRIPTION
**Pull request by IgnatovFedor**

### CODE REVIEW

File: poller.py [hunk 1]
```diff
@@ -155,15 +166,15 @@ def _process(self, payload: dict) -> None:
 def main() -> None:
     args = parser.parse_args()
 
-    pipeline_config_path = find_config(args.config_path)
+    model_url = args.model_url
```
Comment (litinsky) [inline poller.py:165]: There should be possibility to set `model_url` both in command line args and poller config file

**Review by litinsky — CHANGES_REQUESTED** [review]

File: tests/poller_tests.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    def __init__(self, tests_keeper):
+        self.result = []
+        self.messages_to_process = 0
+        self.all_messages_processed = asyncio.Event()
+        self.all_messages_processed.set()
+        self._test_samples = tests_keeper.get_tests()
+        self._test_configs = tests_keeper.get_configs()
+
+        # Start poller
+        class Crutch:
```
Comment (litinsky) [inline tests/poller_tests.py]: What is it for?

File: poller.py (hunk 1 above)
Comment (IgnatovFedor) [reply poller.py:165]: Done

**Review by litinsky — CHANGES_REQUESTED** [review]

File: tests/poller_tests.py (hunk 2 above)
Comment (IgnatovFedor) [reply tests/poller_tests.py]: Crutch class is for replacing argument parser in poller without rewriting poller

File: poller.py [hunk 3]
```diff
@@ -1,24 +1,17 @@
...
-from deeppavlov.core.commands.infer import build_model
-from deeppavlov.core.common.file import read_json
-from deeppavlov.deep import find_config
-from deeppavlov.download import deep_download
-
+from multiprocessing import Process, Queue
+from pathlib import Path
+from typing import Optional
 
 logging.disable(logging.DEBUG)
```
Comment (litinsky) [inline poller.py]: You should not disable logging level in code. Please parametrise logging level in config and set corresponding level, not disable unnecessary.

File: tests/poller_tests.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
+import asyncio
+import json
+import logging
+import numpy as np
+import poller
```
Comment (litinsky) [inline tests/poller_tests.py]: Please feat import order according PEP8

File: tests/poller_tests.py (hunk 2 above)
Comment (IgnatovFedor) [reply tests/poller_tests.py]: Crutch removed. Wrapper class now is testing directly

File: poller.py [hunk 5]
```diff
@@ -73,50 +59,68 @@ def _process_input(self, input_q: dict) -> None:
...
-                batch = list(zip(*chunk))
-                ids_batch = list(batch[0])
-                utts_batch = list(batch[1])
-                response_batch = self.agent(utts_batch, ids_batch)
-
-                for resp in zip(ids_batch, response_batch):
-                    self._send_results(*resp)
+        tasks = (self.loop.create_task(self._process_chats_batch(chats_batch,
+                                                                 msg_id,
+                                                                 chat_ids)) for msg_id, chats_batch in enumerate(batched_chats))
```
Comment (litinsky) [inline poller.py]: Here and further naming: `msg_id` -> `layer_id`

File: poller.py [hunk 6]
```diff
@@ -73,50 +59,68 @@ def _process_input(self, input_q: dict) -> None:
...
+        utts_batch = [(chat_ids[utt_id], utt) for utt_id, utt in enumerate(chats_batch) if utt]
+        j = self.config['infer_batch_length']
+        chunked_utts_batch = [utts_batch[i * j:(i + 1) * j] for i in range((len(utts_batch) + j - 1) // j)]
+        await asyncio.gather(*(self.loop.create_task(self._process_chunk(chunk, msg_id)) for chunk in chunked_utts_batch))
+
+    async def _process_chunk(self, chunk: list, msg_id: int) -> None:
+        batch = list(zip(*chunk))
+        ids_batch = list(batch[0])
+        utts_batch = list(batch[1])
+        data = {"text1": utts_batch}
```
Comment (litinsky) [inline poller.py]: Parametrize: `"text1"` in `{"text1": utts_batch}`

File: poller.py [hunk 7]
```diff
@@ -73,50 +59,68 @@ def _process_input(self, input_q: dict) -> None:
...
+        try:
+            response = await self.loop.run_in_executor(None, functools.partial(requests.post,
+                                                                               self.config['model_url'],
+                                                                               json=data,
+                                                                               timeout=self.config['request_timeout']))
+        except requests.exceptions.ReadTimeout:
+            response = requests.Response()
+            response.status_code = 503
+        if response.status_code == 200:
+            tasks = (self.loop.create_task(self._send_results(*resp, msg_id)) for resp in zip(ids_batch, response.json()))
```
Comment (litinsky) [inline poller.py]: `self._send_results(*resp, msg_id)` - > ``self._send_results(resp[0], resp[1], msg_id)``

File: poller.py [hunk 8]
```diff
@@ -73,50 +59,68 @@ def _process_input(self, input_q: dict) -> None:
...
-                response_batch = self.agent(utts_batch, ids_batch)
-
-                for resp in zip(ids_batch, response_batch):
-                    self._send_results(*resp)
+        tasks = (self.loop.create_task(self._process_chats_batch(chats_batch,
+                                                                 msg_id,
+                                                                 chat_ids)) for msg_id, chats_batch in enumerate(batched_chats))
+        await asyncio.gather(*tasks)
+
+    async def _process_chats_batch(self, chats_batch: list, msg_id: int, chat_ids: list) -> None:
```
Comment (litinsky) [inline poller.py]: Here, before and further use type annotation from `typing` if possible: for example `List[int]` or `Dict[str, List[str]]`.

File: poller.py [hunk 9]
```diff
@@ -73,50 +59,68 @@ def _process_input(self, input_q: dict) -> None:
...
+        utts_batch = [(chat_ids[utt_id], utt) for utt_id, utt in enumerate(chats_batch) if utt]
+        j = self.config['infer_batch_length']
+        chunked_utts_batch = [utts_batch[i * j:(i + 1) * j] for i in range((len(utts_batch) + j - 1) // j)]
+        await asyncio.gather(*(self.loop.create_task(self._process_chunk(chunk, msg_id)) for chunk in chunked_utts_batch))
+
+    async def _process_chunk(self, chunk: list, msg_id: int) -> None:
+        batch = list(zip(*chunk))
```
Comment (litinsky) [inline poller.py]: Naming `batch` `->` `ids_utts_batch`.

File: poller.py [hunk 10]
```diff
@@ -73,50 +59,68 @@ def _process_input(self, input_q: dict) -> None:
...
+        data = {"text1": utts_batch}
+        try:
+            response = await self.loop.run_in_executor(None, functools.partial(requests.post,
+                                                                               self.config['model_url'],
+                                                                               json=data,
+                                                                               timeout=self.config['request_timeout']))
+        except requests.exceptions.ReadTimeout:
+            response = requests.Response()
+            response.status_code = 503
+        if response.status_code == 200:
```
Comment (litinsky) [inline poller.py:97]:
```suggestion
        
        if response.status_code == 200:
```

File: poller.py [hunk 11]
```diff
@@ -73,50 +59,68 @@ def _process_input(self, input_q: dict) -> None:
...
+            tasks = (self.loop.create_task(self._send_results(*resp, msg_id)) for resp in zip(ids_batch, response.json()))
+        else:
+            self.log.error(f'Got {response.status_code} code from {self.config["model_url"]}')
+            tasks = (self.loop.create_task(self._send_results(*resp, msg_id)) for resp in zip_longest(ids_batch, [], fillvalue='Server error'))
+        await asyncio.gather(*tasks)
```
Comment (litinsky) [inline poller.py:104]:
```suggestion
        
        await asyncio.gather(*tasks)
```

File: poller.py (hunk 7 above)
Comment (litinsky) [reply poller.py]: It is better to rewrite it in more explicit style.

File: poller.py [hunk 12]
```diff
@@ -73,50 +59,68 @@ def _process_input(self, input_q: dict) -> None:
...
+                                                                               json=data,
+                                                                               timeout=self.config['request_timeout']))
+        except requests.exceptions.ReadTimeout:
+            response = requests.Response()
+            response.status_code = 503
+        if response.status_code == 200:
+            tasks = (self.loop.create_task(self._send_results(*resp, msg_id)) for resp in zip(ids_batch, response.json()))
+        else:
+            self.log.error(f'Got {response.status_code} code from {self.config["model_url"]}')
+            tasks = (self.loop.create_task(self._send_results(*resp, msg_id)) for resp in zip_longest(ids_batch, [], fillvalue='Server error'))
```
Comment (litinsky) [inline poller.py]: May be tasks tuple should be composed once, after all checks.

**Review by litinsky — CHANGES_REQUESTED** [review]

File: poller.py (hunk 11 above)
Comment (IgnatovFedor) [reply poller.py:104]: Done

File: poller.py (hunk 10 above)
Comment (IgnatovFedor) [reply poller.py:97]: Done

File: poller.py (hunk 8 above)
Comment (IgnatovFedor) [reply poller.py]: Done for lists. Dicts left unchanged due to heterogeneous content.

File: poller.py [hunk 13]
```diff
@@ -73,50 +59,71 @@ def _process_input(self, input_q: dict) -> None:
...
-
-        return processed_message
-
-    def _send_results(self, chat_id, response) -> None:
-        resp_text = str("{\"text\":\"" + str(response) + "\"}")
+            message = payload['text']
+        return message
 
+    async def _send_results(self, chat_id: int, response: list, layer_id: int) -> None:
+        buf = {'text': ' '.join(str(element) for element in response)}
```
Comment (IgnatovFedor) [inline poller.py:115]: If model's response is `['']` (for example ODQA could send such response) then resp_text (see below) equals `{'text': ''}`. Router bot not sending any message to user with empty `text` field. What better to send in such cases? Maybe `-` to give user information that model respond nothing and not just ignored last message?

Comment (litinsky) [reply poller.py:115]: Not need for now, I think, we should simply translate model answers until router-bot user say to change this behavior.

**Review by litinsky — APPROVED** [review]

**Pull request merged**